File: language/src/main/java/brj/nodes/ValueExprRootNode.java
package brj.nodes;

import brj.AnalyserKt;
import brj.BridjeLanguage;
import com.oracle.truffle.api.dsl.NodeChild;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.FrameDescriptor;
import com.oracle.truffle.api.nodes.RootNode;

@NodeChild(value = "writeFxLocal", type = WriteLocalNode.class)
@NodeChild(value = "expr", type = ExprNode.class)
public abstract class ValueExprRootNode extends RootNode {
    protected ValueExprRootNode(BridjeLanguage language, FrameDescriptor frameDescriptor) {
        super(language, frameDescriptor);
    }

    @Specialization
    public Object execute(Object _writeFxLocal, Object expr) {
        return expr;
    }

    public static ValueExprRootNode create(BridjeLanguage lang, FrameDescriptor frameDescriptor, ExprNode exprNode) {
        WriteLocalNode readFxMapNode = WriteLocalNodeGen.create(
            new ReadArgNode(lang, 0),
            frameDescriptor.findOrAddFrameSlot(AnalyserKt.getDEFAULT_FX_LOCAL()));
        return ValueExprRootNodeGen.create(lang, frameDescriptor, readFxMapNode, exprNode);
    }
}

File: language/src/main/kotlin/brj/nodes/ValueExprRootNode.kt
package brj.nodes

import brj.BridjeLanguage
import brj.DEFAULT_FX_LOCAL
import brj.nodes.ExprNode
import com.oracle.truffle.api.dsl.NodeChild
import com.oracle.truffle.api.dsl.NodeChildren
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.frame.FrameDescriptor
import com.oracle.truffle.api.nodes.RootNode

@NodeChildren(
    NodeChild(value = "writeFxLocal", type = WriteLocalNode::class),
    NodeChild(value = "expr", type = ExprNode::class)
)
abstract class ValueExprRootNode protected constructor(lang: BridjeLanguage, frameDescriptor: FrameDescriptor) :
    RootNode(lang, frameDescriptor) {
    @Specialization
    fun execute(@Suppress("UNUSED_PARAMETER") _writeFxLocal: Any, expr: Any) = expr

    companion object {
        fun create(lang: BridjeLanguage, frameDescriptor: FrameDescriptor, exprNode: ExprNode): ValueExprRootNode {
            val readFxMapNode = WriteLocalNodeGen.create(
                lang,
                ReadArgNode(lang, 0),
                frameDescriptor.findOrAddAuxiliarySlot(DEFAULT_FX_LOCAL)
            )
            return ValueExprRootNodeGen.create(lang, frameDescriptor, readFxMapNode, exprNode)
        }
    }
}